openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -0,0 +1,255 @@
+            setting_name (String): The name of the flag or switch.
+
+        """
+        return u'{}.{}'.format(self.namespace, setting_name)
+
+    @staticmethod
+    def _get_namespace_request_cache():
+        """
+        Returns the request cache used by WaffleNamespace classes.
+        """
+        return get_request_cache('WaffleNamespace')
+
+
+class WaffleSwitchNamespace(WaffleNamespace):
+    """
+    Provides a request cached namespace for waffle switches.
+    """
+    def is_enabled(self, switch_name):
+        """
+        Returns and caches whether the given waffle switch is enabled.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        value = self._cached_switches.get(namespaced_switch_name)
+        if value is None:
+            value = switch_is_active(namespaced_switch_name)
+            self._cached_switches[namespaced_switch_name] = value
+        return value
+
+    @contextmanager
+    def override(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the duration of this
+        contextmanager.
+        Note: The value is overridden in the request cache AND in the model.
+        """
+        previous_active = self.is_enabled(switch_name)
+        try:
+            self.override_for_request(switch_name, active)
+            with self.override_in_model(switch_name, active):
+                yield
+        finally:
+            self.override_for_request(switch_name, previous_active)
+
+    def override_for_request(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the remainder of
+        this request (as this is not a context manager).
+        Note: The value is overridden in the request cache, not in the model.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        self._cached_switches[namespaced_switch_name] = active
+        log.info(u"%sSwitch '%s' set to %s for request.", self.log_prefix, namespaced_switch_name, active)
+
+    @contextmanager
+    def override_in_model(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the duration of this
+        contextmanager.
+        Note: The value is overridden in the model, not the request cache.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        with waffle_override_switch(namespaced_switch_name, active):
+            log.info(u"%sSwitch '%s' set to %s in model.", self.log_prefix, namespaced_switch_name, active)
+            yield
+
+    @property
+    def _cached_switches(self):
+        """
+        Returns a dictionary of all namespaced switches in the request cache.
+        """
+        return self._get_namespace_request_cache().setdefault('switches', {})
+
+
+class WaffleFlagNamespace(WaffleNamespace):
+    """
+    Provides a request cached namespace for waffle flags.
```

Comment:
This isn't a namespace for Waffle flags, to my mind, since it implements the Waffle flag interface. It seems like a flag that additionally has a namespace.